src/analytics/ledger/output_activity.rs

```diff
@@ -0,0 +1,185 @@
+                if let Output::Nft(nft_output) = &ledger_output.output {
+                    if nft_output.nft_id == NftId::implicit() {
+                        // Convert implicit ids to explicit ids to make all nfts comparable
+                        Some(NftId::from(ledger_output.output_id))
+                    } else {
+                        Some(nft_output.nft_id)
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect::<HashSet<_>>();
+
+        self.measurement.created_count += nft_outputs.difference(&nft_inputs).count() as u64;
+        self.measurement.transferred_count += nft_outputs.intersection(&nft_inputs).count() as u64;
+        self.measurement.destroyed_count += nft_inputs.difference(&nft_outputs).count() as u64;
+    }
+
+    fn end_milestone(&mut self, _: MilestoneIndexTimestamp) -> Option<Self::Measurement> {
+        Some(self.measurement)
+    }
+}
+
+/// Alias activity statistics.
+#[derive(Copy, Clone, Debug, Default, PartialEq, AddAssign, SubAssign)]
+pub struct AliasActivityMeasurement {
+    pub created_count: u64,
+    pub governor_changed_count: u64,
+    pub state_changed_count: u64,
+    pub destroyed_count: u64,
+}
+
+/// Measures the ledger Alias activity.
+pub struct AliasActivityAnalytics {
+    measurement: AliasActivityMeasurement,
+}
+
+struct AliasData {
+    alias_id: AliasId,
+    governor_address: Address,
+    state_index: u32,
+}
+
+impl std::cmp::PartialEq for AliasData {
+    fn eq(&self, other: &Self) -> bool {
+        self.alias_id == other.alias_id
+    }
+}
+
+impl std::cmp::Eq for AliasData {}
+
+impl std::hash::Hash for AliasData {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.alias_id.hash(state);
+    }
+}
+
+impl TransactionAnalytics for AliasActivityAnalytics {
+    type Measurement = AliasActivityMeasurement;
+
+    fn begin_milestone(&mut self, _: MilestoneIndexTimestamp) {}
+
+    fn handle_transaction(&mut self, inputs: &[LedgerSpent], outputs: &[LedgerOutput]) {
+        let alias_inputs = inputs
+            .iter()
+            .filter_map(|ledger_spent| {
+                if let Output::Alias(alias_output) = &ledger_spent.output.output {
+                    let alias_id = if alias_output.alias_id == AliasId::implicit() {
+                        // Convert implicit ids to explicit ids to make all aliases comparable
+                        AliasId::from(ledger_spent.output.output_id)
+                    } else {
+                        alias_output.alias_id
+                    };
+                    Some(AliasData {
+                        alias_id,
+                        governor_address: alias_output.governor_address_unlock_condition.address,
+                        state_index: alias_output.state_index,
+                    })
+                } else {
+                    None
+                }
+            })
+            .collect::<HashSet<_>>();
```

Comment:
We should use `HashMap<*Id,(Address,StateIndex)>` for `alias_inputs` and `alias_outputs`. To get the intersection we just iterate over either of them and index into the second. This would allow us to get rid of the ad-hoc implementations of `hash`, as well as the `AliasData` struct.

Reply:
Do we want to ship or do we want to make things perfect out-of-the-gate? This is an implementation detail that can be changed later without breaking anything. Are you sure you want this changed?

Reply:
That's the thing, it's not an implementation detail, because the trait implementations are in the public interface.

Reply:
I don't understand. The way the computation itself is done is not public. `AliasData` is private.

Reply:
@Alexandcoats any thoughts?

Reply:
Am I wrong or did Jochen cede the point here?

Reply:
yup, I didn't see the edit in time.